src/mlpack/core/optimizers/fw/update_span.hpp

```diff
@@ -0,0 +1,144 @@
+/**
+ * @file update_span.hpp
+ * @author Chenzhe Diao
+ *
+ * Update method for FrankWolfe algorithm, recalculate the optimal in the span
+ * of previous solution space. Used as UpdateRuleType.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_FW_UPDATE_SPAN_HPP
+#define MLPACK_CORE_OPTIMIZERS_FW_UPDATE_SPAN_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Recalculate the optimal solution in the span of all previous solution space, 
+ * used as update step for FrankWolfe algorithm.
+ *
+ *
+ * For UpdateSpan to work, FunctionType template parameters are required.
+ * This class must implement the following functions:
+ *
+ * FunctionType:
+ *
+ *   double Evaluate(const arma::mat& coordinates);
+ *   void Gradient(const arma::mat& coordinates,
+ *                 arma::mat& gradient);
+ *
+ *
+ * @tparam FunctionType Objective function type to be minimized in FrankWolfe algorithm.
+ */
+template<typename FunctionType>
+class UpdateSpan
+{
+ public:
+  /**
+   * Construct the span update rule. The function to be optimized is 
+   * input here.
+   *
+   * @param function Function to be optimized in FrankWolfe algorithm.
+   */
+  UpdateSpan(FunctionType& function): function(function)
+  { /* Do nothing. */ }
+
+ /**
+  * Update rule for FrankWolfe, reoptimize in the span of original solution space.
+  *
+  *
+  * @param old_coords previous solution coords.
+  * @param s current linear_constr_solution result.
+  * @param new_coords new output solution coords.
+  * @param num_iter current iteration number
+  */
+  void Update(const arma::mat& old_coords,
+      const arma::mat& s,
+      arma::mat& new_coords,
+      const size_t num_iter)
+  {
+      // add atom here
+      arma::uvec ind = find(s, 1);
+      arma::uword d = ind(0);
+      AddAtom(d);
+
+      arma::vec b = function.Vectorb();
+      arma::mat x = solve(atoms_current, b);
```

Comment:
This might be a design problem.

The original FrankWolfe solves "constrained" optimization problem, while this class is supposed to update the solution from the space spanned by all previous solutions, which is unconstrained.  So theoretically it won't use this `UpdateSpan` class. This class is used to solve the unconstrained problem with sparse solution, such as OMP.

So the type of problems that would use this method I currently have in mind are just the atom domain optimization problems, that is, find sparse solution to ||y-Ax||^2. So in vector case, it is just the `FuncSq` class, while I am also implementing a matrix case for matrix completion problem.

In summary, I think we don't need the `FunctionType` interface in `FrankWolfe` class to have `MatrixA()` and `Vectorb()` function, but we can require the `FunctionType` interface in `UpdateSpan` class to have them. I think that will make more sense. I can add them in the documentation.